### PR TITLE
Added lambda expressions and function object in GDScript.

### PR DESCRIPTION
**Pull request by dbsGen**
```
func _ready():

  # Easier connection
  self.connect("renamed", on_renamed)

  var hello = "hello"
  # lambda expressions
  var lambda_func = func():
      print(hello)

  # or 
  func lambda_func2():
      print(hello)

  # lambda_func is a function object
  lambda_func.apply()
  lambda_func2.apply()

  # A native function object
  var native_func = self.set_name
  native_func.apply("new_name")


func on_renamed():
  print(get_name())
```

### CODE REVIEW

**Comment by akien-mga** [comment]
Would address the feature request #2684 if/when it's merged.

**Comment by dbsGen** [comment]
@akien-mga Cool.

**Comment by reduz** [comment]
wow, amazing, will give it a check next week

**Comment by KillerJaguar** [comment]
Really neat! I did some tests and it seems to be working fine. It captures variables properly and keeps proper scope. You can even use them as function arguments, though the white spacing is a little finicky.

```
func _ready():

	for_each([1,2,3,4,5], func(i): 
		print(i)
	, func(i):
		return i * i
	)

func for_each(array, fn, fn2):
	for i in array:
		fn.apply(fn2.apply(i))
```

It would be nicer if you could just do fn() instead of fn.apply().

**Comment by vnen** [comment]
Citing myself from #2684:
> As one note, the first argument of `lambda.apply()` should be the `self` object, so `self` references inside the lambda are properly resolved (or a `lambda.bind(object)` should be provided).

**Comment by dbsGen** [comment]
@KillerJaguar I am working on it. Now it run fine on my computer, but there are some bugs.

**Comment by dbsGen** [comment]
@vnen Good idea. But it is a bit difficult. I am thinking about how to implement。

**Comment by dbsGen** [comment]
@vnen At last choose a easy way to implement it.

**Comment by dbsGen** [comment]
@reduz I want to implement this feature:

     func _ready():
        block var v1 = 0
        func f():
          v1 += 1
        f()
        print(v1)

In a lambda function you have ability to change the value which out of the lambda function.
In my mind, to implement this feature I have to add a Reference Variant class to core and add a new type to Variant types.

I want to ask is this feature nessesary? Now we can do that by this way.

     func _ready():
        var obj = C1.new()
        func f():
          obj.v1 += 1
        f()
        print(obj.v1)

      class C1:
        var v1 = 0

Or other better way to implement it?

**Comment by reduz** [comment]
This PR is huge, will take my time reading it. My main concern is how much this new feature can impact performance. Wil have to test it..

**Comment by dbsGen** [comment]
@reduz Get it, I do not implement it this time. It will be a new PR, after this PR is OK.

**Comment by akien-mga** [comment]
@reduz Should we merge it in a feature branch for the time being, so that @dbsGen can continue working on it via other PRs, and we can assess the performance impact before merging into master?

**Comment by reduz** [comment]
I would like to give it a check first, didn't have enough time to do it but
hopefully will this week

**Comment by dbsGen** [comment]
OK, it is already marged.

**Comment by dbsGen** [comment]
Rebased

**Comment by jackmakesthings** [comment]
Awesome, can't wait to see this in action.

**Comment by reduz** [comment]
Going to do some changes to GDScript soon so holding on this one for some
more time

**Comment by TX256** [comment]
I like GDScript, but I can't help but to miss some kind of simple "lambdas/callbacks" when using signals.

Current code:

```
func animate():
	var tween = Tween.new()
	add_child(tween)
	tween.animate_property(...)
	tween.connect("tween_complete", self, "_tween_complete", [tween])
	tween.start()

func _tween_complete(object, key, tween):
	tween.queue_free()
```

I'd love to be able to do something like this:

```
func animate():
	var tween = Tween.new()
	add_child(tween)
	tween.animate_property(...)
	tween.on("tween_complete") -> [tween]:
		tween.queue_free()
	tween.start()

```

**Comment by reduz** [comment]
you can use yield for this currently:

tween.start()
yield(tween,"tween_complete")
tween.queue_free()

**Comment by TX256** [comment]
Black magic! That didn't even cross my mind, as GDScript happens to be my first encounter with co-routines - and I'm so rooted to deal with these kinds of callbacks with lambdas in some other languages.

**Comment by reduz** [comment]
still intend to merge this at some point, need to give it a through check

**Comment by Zylann** [comment]
@reduz I just searched the issues and found this PR. Please make it happen, as I need lambdas so badly at this exact moment :D

**Comment by dbsGen** [comment]
Rebased, I had run some test code and it works for me, maybe need some more test.

**Comment by vnen** [comment]
@dbsGen this is not properly rebased, you added merge commits and there are still conflicts. I know this has been waiting for review a long time but in the current state is pretty much impossible to do so (look a the lines changed number: 1,381,131 lines added, 1,197,704 lines removed in 8543 files, it's just absurd).

I can provide some help if you need to rebase this properly. You need to remove the merge commits and reapply your changes on top of the master branch (then force push) so the conflicts go away and only your changes are shown in this PR.

**Comment by dbsGen** [comment]
My mistake, it is fixed now.

**Comment by vnen** [comment]
Much better, thanks! I hope @reduz get some time to look at this, it'd be a nice addition.

File: modules/gdscript/gd_parser.h [hunk 1]
```diff
@@ -141,7 +141,8 @@ class GDParser {
 			if (body->variables.find(p_key) < 0 && body->arguments.find(p_key) < 0) {
 				if (require_keys.find(p_key) < 0) {
 					require_keys.push_back(p_key);
-				}else if (LambdaFunctionNode *func = dynamic_cast<LambdaFunctionNode*>(parent)) {
+				}else if (parent->type == Node::TYPE_LAMBDA_FUNCTION) {
+					LambdaFunctionNode *func = static_cast<LambdaFunctionNode*>(parent)
```
Comment (vnen) [inline modules/gdscript/gd_parser.h]: Missed a `;` here.

**Comment by bojidar-bg** [comment]
That's going to need a huge squash, as akien might be concerned :wink:

**Comment by zatherz** [comment]
Bump. Amazing PR.

**Comment by razvanc-r** [comment]
I personally don't like the syntax of this lambda function. Is it limited to only one statement? If yes I'd definitely go for something like what Python has, if not I would rethink it

**Comment by karroffel** [comment]
@razvanc-r it's not limited to one statement, it's using `_parse_function` that's used for top-level functions and functions in subclasses too, so it parses full functions with a block of code, not just one expression.

Also a rebase is needed @dbsGen, I'd like to review but it's not compatible atm

**Comment by razvanc-r** [comment]
@karroffel well I'm not interested in the internals :P, I'm just giving my feedback. I think a different syntax should be used for multi-line lambda functions. That is all

**Comment by razvanc-r** [comment]
Wait, does this PR also change how `Object.connect` works? I personally prefer the way it's implemented now: `self.connect('signal', object, function_name)` instead of `self.connect('signal', function_name)`. Because I think explicit is better than implicit.

Also, is there no way to just call the lambda function like normal functions (`lambda()` instead of `lambda.apply()`)? I know it's an object in the end but I think it should still be called just like regular functions.

**Comment by zatherz** [comment]
@razvanc-r I'm pretty sure this commit https://github.com/godotengine/godot/pull/2704/commits/a7fa288586aec72e0ac6a78d045ef147b31d40d6 makes it possible to call them directly

**Comment by Zylann** [comment]
Also, why "apply"? something like "call" or "call_func" looks more appropriate for a callable object, though I like being able to use () directly as well.

**Comment by vnen** [comment]
@Zylann 
> Also, why "apply"?  something like "call" or "call_func" looks more appropriate for a callable object

I would guess this comes from Javascript, See my comment earlier: https://github.com/godotengine/godot/pull/2704#issuecomment-152748116. Though if the functionality steers away from JS, there's no reason to name it `apply()`.

**Comment by gilch** [comment]
It might be worth comparing to [Boo's anonymous function syntax](https://github.com/boo-lang/boo/wiki/Closures). Boo also has indentation-based block delimiters like GDScript and Python.

**Comment by toger5** [comment]
NICE. just found that pr and love it. Always was thinking about creating a module for nice swift style api calls. (something like alamofire ;) )
With this it would be soon easy to do a http wrapper. (current http in gdscript is still not that elegant)

would it be possible to add some syntax sugar and even remove the func keyword when passing functions?

```
#so that you can leave away the func
callMultipleTimes(10, (iteration, total):
    print("called this function for the " + iteration + "time")
    print("will call it " + (total - iteration - 1) + "more times")
)

callMultipleTimes(times, function):
   for i in range(times):
        function(i,times)


#with current it would look like this if I'm not mistaking:
callMultipleTimes(10, func(iteration, total):
    print("called this function for the " + iteration + "time")
    print("will call it " + (total - iteration - 1) + "more times")
)

```

**Comment by toger5** [comment]
and there definitely should be an implementation like this
```
myObj.connect("someSignal", ():
    myObj.position.x += 10
)
```
this would be soooo beautiful!!!

**Comment by dbsGen** [comment]
Merged, I had tested  those code.

	extends Node

	# class member variables go here, for example:
	# var a = 2
	# var b = "textvar"

	func _ready():
		# Easier connection
		self.connect("renamed", on_renamed)

		var hello = "hello"
		# lambda expressions
		var lambda_func = func():
			print(hello)
			print("f1")

		# or 
		func lambda_func2():
			print(hello)
		
		call_func(func(v):
			print("call:" + str(v))
		, 123
		
		)
		# lambda_func is a function object
		lambda_func()
		lambda_func2()

		# A native function object
		var native_func = self.set_name
		native_func("new_name")

	func call_func(f, arg):
		print("call func:")
		f(arg)

	func on_renamed():
		print(get_name())

**Comment by akien-mga** [comment]
As this PR has been around for a while, rebased many times (thanks @dbsGen!) and I'm unsure that @reduz would merge it as is in the near future, I'm thinking that it would likely be worth it to merge it manually into a feature branch of the @godotengine repo, e.g. `godotengine/lambda`.

This would make it easier to test for interested Godot users, and to rebase when there's a need as all Godot devs would have commit rights on the branch (and other contributors could do PRs if they want).

It would also make it easier for me (or any interested dev) to "rebase" this work once the clang-format reformat commit has been applied (which will break all PRs). https://engineering.mongodb.com/post/succeeding-with-clangformat-part-3-persisting-the-change/ describes a nice workflow to handle this. To avoid hassle, I'd probably squash all commits of this PR into one huge commit.

WDYT?

**Comment by dbsGen** [comment]
@akien-mga It sounds good , what should I do?

**Comment by akien-mga** [comment]
@dbsGen Nothing, I'll do it myself manually this evening, based on this PR.

**Comment by akien-mga** [comment]
It's now all squashed and pushed in the `lambda` branch: https://github.com/godotengine/godot/tree/lambda
Please test :)

Closing the PR as it's superseded by that branch.

**Comment by zatherz** [comment]
YES, FINALLY!

**Comment by robertdhernandez** [comment]
Bumping this, because while it has been given its own branch, I feel this has been largely forgotten about since then. Will this get merged into master?

**Comment by reduz** [comment]
Probably not in the short time, Improvements to GDScript will wait until
after 3.0

On Sat, Jul 15, 2017 at 10:32 AM, Robert Hernandez <notifications@github.com

**Comment by zatherz** [comment]
Why? Is it too late? If so then unfortunate, because I'd assume this might come with some backwards incompatible changes, and 3.0 is already expected to be incompatible with 2.0 in a lot of cases.

**Comment by Calinou** [comment]
For future reference, Godot 4.0alpha has first-class Callable and Signal types, in addition to lambdas in GDScript: https://godotengine.org/article/gdscript-progress-report-feature-complete-40

**Pull request closed without merging**